storage/src/index/storage.rs

```diff
@@ -18,63 +24,202 @@ pub struct ValueIterator<'a, V> {
     next: Option<&'a Record<V>>,
 }
 
-// An iterator over all values associated with a translated key, allowing for mutation.
-pub struct MutableValueIterator<'a, V> {
-    next: Option<&'a mut Record<V>>,
-}
-
 impl<V> ValueIterator<'_, V> {
     /// Create a `ValueIterator` that returns no items.
     fn empty() -> Self {
         ValueIterator { next: None }
     }
 }
 
-impl<V> MutableValueIterator<'_, V> {
-    /// Create a `MutableValueIterator` that returns no items.
-    fn empty() -> Self {
-        MutableValueIterator { next: None }
-    }
-}
-
 impl<'a, V> Iterator for ValueIterator<'a, V> {
     type Item = &'a V;
 
     fn next(&mut self) -> Option<Self::Item> {
         match self.next {
             Some(next) => {
                 let loc = &next.value;
-                self.next = next.next.as_deref();
+                match next.next {
+                    Some(ref next_next) => self.next = Some(next_next),
+                    None => self.next = None,
+                }
+                Some(loc)
+            }
+            None => None,
+        }
+    }
+}
+
+/// A wrapper for the hashmap entry when it can be either occupied or vacant.
+enum OccupiedOrVacant<'a, K, V> {
```

Comment:
I suggest just calling this `Entry`

Reply:
Changed to `MapEntry` so it doesn't conflict with `hash_map::Entry`